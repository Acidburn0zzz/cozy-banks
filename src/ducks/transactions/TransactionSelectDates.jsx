import React from 'react'
import SelectDates from 'components/SelectDates'
import uniq from 'lodash/uniq'
import { subMonths, format, parse, differenceInCalendarMonths } from 'date-fns'

const rangeMonth = (startDate, endDate) => {
  const options = []

  for (let i = 0; i < differenceInCalendarMonths(endDate, startDate); i++) {
    options.push(subMonths(endDate, i))
  }

  return options
}

const getYearMonth = transaction => transaction.date.slice(0, 7)

export const getOptions = transactions => {
  const availableMonths = uniq(transactions.map(getYearMonth)).sort()

  const mAvailableMonths = new Set(availableMonths)

  const start = parse(availableMonths[0], 'YYYY-MM')
  const end = parse(availableMonths[availableMonths.length - 1], 'YYYY-MM')

  return rangeMonth(start, end).map(month => {
    const fmted = format(month, 'YYYY-MM')
    return {
      yearMonth: fmted,
      disabled: !mAvailableMonths.has(fmted)
    }
  })
}

const TransactionSelectDates = ({ onChange, transactions, value }) => {
  return (
    <SelectDates
      onChange={onChange}
      options={getOptions(transactions)}
      value={value}
    />
  )
}

export default TransactionSelectDates
