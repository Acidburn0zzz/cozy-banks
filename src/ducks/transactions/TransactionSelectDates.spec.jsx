import { getOptions } from './TransactionSelectDates'
import fixtures from 'test/fixtures'

const transactions = fixtures['io.cozy.bank.operations']


describe('options from select dates', () => {
  it('should compute correctly', () => {
    expect(getOptions(transactions)).toMatchSnapshot()
  })
})
