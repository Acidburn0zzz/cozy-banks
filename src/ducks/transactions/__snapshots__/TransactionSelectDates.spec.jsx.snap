// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options from select dates should compute correctly 1`] = `
Array [
  Object {
    "disabled": false,
    "yearMonth": "2018-06",
  },
  Object {
    "disabled": true,
    "yearMonth": "2018-05",
  },
  Object {
    "disabled": true,
    "yearMonth": "2018-04",
  },
  Object {
    "disabled": true,
    "yearMonth": "2018-03",
  },
  Object {
    "disabled": true,
    "yearMonth": "2018-02",
  },
  Object {
    "disabled": false,
    "yearMonth": "2018-01",
  },
  Object {
    "disabled": true,
    "yearMonth": "2017-12",
  },
  Object {
    "disabled": true,
    "yearMonth": "2017-11",
  },
  Object {
    "disabled": true,
    "yearMonth": "2017-10",
  },
  Object {
    "disabled": true,
    "yearMonth": "2017-09",
  },
  Object {
    "disabled": false,
    "yearMonth": "2017-08",
  },
  Object {
    "disabled": false,
    "yearMonth": "2017-07",
  },
  Object {
    "disabled": false,
    "yearMonth": "2017-06",
  },
]
`;
